doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -444,34 +457,13 @@ the conversion in hypotheses :n:`{+ @ident}`.
    the call-by-value strategy, the latter is generally more efficient for
    evaluating purely computational expressions (i.e. with little dead code).
 
-.. tacv:: compute
-          cbv
-   :name: compute; _
-
-   These are synonyms for ``cbv beta delta iota zeta``.
-
-.. tacv:: lazy
-
-   This is a synonym for ``lazy beta delta iota zeta``.
-
-.. tacv:: compute [ {+ @qualid} ]
-          cbv [ {+ @qualid} ]
-
-   These are synonyms of :n:`cbv beta delta {+ @qualid} iota zeta`.
-
-.. tacv:: compute - [ {+ @qualid} ]
-          cbv - [ {+ @qualid} ]
+.. tacn:: compute {? @delta_flag } {? @occurrences }
 
-   These are synonyms of :n:`cbv beta delta -{+ @qualid} iota zeta`.
+   .. todo PR: right (below)?  If so, let's drop one of these tactics.  Or put with cbv above
```

Comment:
We can't say `compute` is a synonym when it has different syntax.  I'll fix that.

A superficial examination of the grammar raises some additional consistency questions:

- Should `compute` be more similar to `vm_compute` and `native_compute`?
- Why does `red_expr` have occurrences for some operations but not others?
- Why are some `[ reference_occs | pattern_occs ]` while others are just `occurrences`?
- Why do `vm_compute` and `native_compute` have both `[ reference_occs | pattern_occs ]` and `occurrences`?

If it makes sense to harmonize these, I guess we should open an issue.

```
red_expr: [
| "red"
| "hnf"
| "simpl" OPT delta_flag OPT [ reference_occs | pattern_occs ]
| "cbv" OPT strategy_flag
| "cbn" OPT strategy_flag
| "lazy" OPT strategy_flag
| "compute" OPT delta_flag
| "vm_compute" OPT [ reference_occs | pattern_occs ]
| "native_compute" OPT [ reference_occs | pattern_occs ]
  :

simple_tactic: [
| "red" OPT occurrences
| "hnf" OPT occurrences
| "simpl" OPT delta_flag OPT [ reference_occs | pattern_occs ] OPT occurrences
| "cbv" OPT strategy_flag OPT occurrences
| "cbn" OPT strategy_flag OPT occurrences
| "lazy" OPT strategy_flag OPT occurrences
| "compute" OPT delta_flag OPT occurrences
| "vm_compute" OPT [ reference_occs | pattern_occs ] OPT occurrences
| "native_compute" OPT [ reference_occs | pattern_occs ] OPT occurrences

Reply:
> * Should `compute` be more similar to `vm_compute` and `native_compute`?

What do you mean?

> * Why does `red_expr` have occurrences for some operations but not others?

Probably historical accident?  It would be nice to have a uniform way to select subterm and perform reduction only in those subterms.

> * Why are some `[ reference_occs | pattern_occs ]` while others are just `occurrences`?

I believe that in fact `occurences` is for selecting whether to reduce in the goal and in which hypotheses, if any, to reduce.  `lazy at 1` gives "Error: Occurrences not supported for this reduction tactic." (`lazy in H` is fine).  The reduction strategies with `[ reference_occs | pattern_occs ]` are the ones that support subterm selection.  And all of the reduction tactics support `occurences` (and none of the `red_expr` do, because it makes no sense to say `eval (lazy in |- * ) in (term)`), and it's just that some of them additionally support `[ reference_occs | pattern_occs ]`.

> * Why do `vm_compute` and `native_compute` have both `[ reference_occs | pattern_occs ]` and `occurrences`?

Having an `at` clause only makes sense if you are selecting a subterm, which `reference_occs` or `pattern_occs` is required for.  Not only `vm_compute` and `native_compute`, though, also `simpl` and `unfold`.

Reply:
I think in fact the desired grammar is
```
red_expr: [
| "red" OPT strategy_flag OPT [ reference_occs | pattern_occs ]
| "hnf" OPT strategy_flag OPT [ reference_occs | pattern_occs ]
| "simpl" OPT strategy_flag OPT [ reference_occs | pattern_occs ]
| "cbv" OPT strategy_flag OPT [ reference_occs | pattern_occs ]
| "cbn" OPT strategy_flag OPT [ reference_occs | pattern_occs ]
| "lazy" OPT strategy_flag OPT [ reference_occs | pattern_occs ]
| "compute" OPT strategy_flag OPT [ reference_occs | pattern_occs ]
| "vm_compute" OPT delta_flag OPT [ reference_occs | pattern_occs ]
| "native_compute" OPT delta_flag OPT [ reference_occs | pattern_occs ]
  :

simple_tactic: [
| "red" OPT strategy_flag OPT [ reference_occs | pattern_occs ] OPT occurrences
| "hnf" OPT strategy_flag OPT [ reference_occs | pattern_occs ] OPT occurrences
| "simpl" OPT strategy_flag OPT [ reference_occs | pattern_occs ] OPT occurrences
| "cbv" OPT strategy_flag OPT [ reference_occs | pattern_occs ] OPT occurrences
| "cbn" OPT strategy_flag OPT [ reference_occs | pattern_occs ] OPT occurrences
| "lazy" OPT strategy_flag OPT [ reference_occs | pattern_occs ] OPT occurrences
| "compute" OPT strategy_flag OPT [ reference_occs | pattern_occs ] OPT occurrences
| "vm_compute" OPT delta_flag OPT [ reference_occs | pattern_occs ] OPT occurrences
| "native_compute" OPT delta_flag OPT [ reference_occs | pattern_occs ] OPT occurrences
```
but this requires significant implementation work to support all of these features.

Reply:
Btw, if you're revisiting names, I expect that it's quite confusing to newcomers that `Compute foo.` means `Eval vm_compute in foo` not `Eval compute in foo` (and you also cannot write `Definition bar := Compute foo.`, which is maybe fine.)

Reply:
I opened #13808 for this--please direct further comments there.

> but this requires significant implementation work to support all of these features.

That's why I made it an issue.

Reply:
> I expect that it's quite confusing to newcomers that `Compute foo.` means `Eval vm_compute in foo` not `Eval compute in foo`

Maybe, but the doc does mention that fact.  Renaming this doesn't see urgent IMHO.